claim_batch/services.py

```diff
@@ -517,27 +519,43 @@ def do_process_batch(audit_user_id, location_id, period, year):
+                .filter(validity_from__lte=end_date)\
+                .filter(validity_from__gte=start_date)\
+                .filter(validity_to__isnull=True)\
+                .filter(process_stamp__lte=end_date)\
+                .filter(process_stamp__gte=start_date)\
+                .filter(Q(items__product__in=products) & Q(items__legacy_id__isnull=True) | Q(services__product__in=products) & Q(services__legacy_id__isnull=True))\
                 .prefetch_related(Prefetch('items', queryset=ClaimItem.objects.filter(legacy_id__isnull=True)))\
                 .prefetch_related(Prefetch('services', queryset=ClaimService.objects.filter(legacy_id__isnull=True)))
+
+            # could be duplicates - distinct
+            claims = list(claims.values_list('id', flat=True).distinct())
+            claims = Claim.objects.filter(id__in=claims)
+
             for claim in claims:
                 remunerated_amount = 0
-                for service in claim.services:
-                    remunerated_amount += service.remunerated_amount
-                for item in claim.items:
-                    remunerated_amount += item.remunerated_amount
+                for service in claim.services.all():
+                    remunerated_amount = service.remunerated_amount + remunerated_amount if service.remunerated_amount else remunerated_amount
+                for item in claim.items.all():
+                    remunerated_amount += item.remunerated_amount + remunerated_amount if item.remunerated_amount else remunerated_amount
+                if remunerated_amount > 0:
+                    claim.valuated = remunerated_amount
+                    claim.save()
+                claim.batch_run = work_data["created_run"]
+                claim.save()
+
             # 5 Generate BatchPayment per product (Ideally per pool but the notion doesn't exist yet)
             # 5.1 filter a calculation valid for batchRun with context BatchPayment (got via 0.2)
-            if work_data.paymentplans:
-                for paymentplan in work_data.paymentplans:
-                    if paymentplan.calculation.active_for_object(batch_run, 'BatchPayment'):
-                        # 54.2 Execute the converter per product/batch run/claim (not claims)
-                        paymentplan.calculation.calculate(work_data, 'BatchPayment'))
+            if work_data["payment_plans"]:
+                for payment_plan in work_data["payment_plans"]:
+                    # 54.2 Execute the converter per product/batch run/claim (not claims)
+                    rcr = run_calculation_rules(payment_plan, "BatchPayment", None,
+                                                work_data=work_data, audit_user_id=audit_user_id,
+                                                location_id=location_id, period=period, year=year)
```

Comment:
the issue with that approach is you limit the data that can be sent, for example the capitation payment needs the allocated contributions

Reply:
I will think how to pass those required arguments necessary to do some calculations etc. Still need to be improved.

Reply:
I was misstaken, you do send work_data as param so it will work fine (whatever is added on the dict will be passed on)

Reply:
ok

Reply:
I think this point could be closed as we determined that my solution is correct